tls/tls_srv.c

```diff
@@ -1287,8 +1287,19 @@ ttls_write_server_key_exchange(TlsCtx *tls, struct sg_table *sgt,
 	 * For suites involving ECDH, extract DH parameters from certificate at
 	 * this point.
 	 */
-	if (ttls_ciphersuite_uses_ecdh(ci))
+	if (ttls_ciphersuite_uses_ecdh(ci)) {
+		ttls_ecdh_init(&TTLS_HS_ecdh_ctx_reset_tag(tls->hs));
+		TTLS_HS_ecdh_ctx(tls->hs).point_format =
+			tls->hs->ecdh_point_format;
 		ttls_get_ecdh_params_from_cert(tls);
+	} else if (ttls_ciphersuite_uses_ecdhe(ci)) {
+		ttls_ecdh_init(&TTLS_HS_ecdh_ctx_reset_tag(tls->hs));
+		TTLS_HS_ecdh_ctx(tls->hs).point_format =
+			tls->hs->ecdh_point_format;
+	} else if (ttls_ciphersuite_uses_dhe(ci)) {
+		ttls_dhm_init(&TTLS_HS_dhm_ctx_reset_tag(tls->hs));
+	}
```

Comment:
This is the place which confuses me: the debugging mess which works in production and if you need to understand the code, you still have no understanding that the `point_format` games are just for debugging.

Reply:
> are just for debugging.

I merged storage spaces of `ecdh_ctx` and `dhm_ctx` into the same union, as they are not used simultaneously. And since they are not separate anymore, we can't just initialize both `ecdh_ctx` and `dhm_ctx` unconditionally. This function is where we decide what we use, and here we are initializing either `dhm_ctx` or `ecdh_ctx`, depending on what is used for the handshake.

Initializations are clearing both `ecdh_ctx` and `dhm_ctx`. If `point_format` is not stored anywhere else, we'll lose it.

One of the solutions will be something like this:
```c
{
    unsigned char pf = tls->hs->ecdh_point_format;
    ttls_ecdh_init(&tls->hs->ecdh_ctx);
    tls->hs->ecdh_point_format = pf;
}
```
But as said in comments above, I'd like to avoid storing anything in `ecdh_ctx` while its memory is used for something else (sha256).

Reply:
Oh, I see, I missed the union for dhm and ecdh.

> I'd like to avoid storing anything in ecdh_ctx while its memory is used for something else (sha256)

Why? If we have a unit test verifying the data structures layouts and sizes to guarantee that they don't overwrite crucial members, will we be good?

BTW https://www.youtube.com/watch?v=GgTwx1399Dc is good to watch - keeping the data structures small is crucial. But we also should not sacrifice the code design clarity.

Reply:
> Why?

That simplifies reasoning about code. With access restrictions no clashing may occur, so we can stop worrying about that and focus on other issues. Checking macros cause immediate crash if access was wrong. So that not only helps to detect existing field access bugs, and also to prevent similar bugs in the future.

> If we have a unit test verifying the data structures layouts and sizes to guarantee that they don't overwrite crucial members, will we be good?

Yeah, kind of. But first we need to find such members, and that's hard to do simply by looking at the code.

As this discussion may continue indefinitely, I propose to extract some changes (merging storage spaces of `ecdh_ctx` and `dhm_ctx`, more precise initialization and freeing of that members, adding explicit `tmp_sha256` field to avoid type casting in the code) into https://github.com/tempesta-tech/tempesta/pull/1254, and close this PR (#1251).